implementations/rust/ockam/ockam_transport_websocket/src/common/router.rs

```diff
@@ -0,0 +1,217 @@
+            trace!("TCP registration request: {} => {}", f, self_addr);
+        } else {
+            // Should not happen
+            return Err(TransportError::InvalidAddress.into());
+        }
+
+        for accept in &accepts {
+            if self.map.contains_key(accept) {
+                return Err(TransportError::AlreadyConnected.into());
+            }
+        }
+        for accept in accepts {
+            let accept_str = accept.to_string();
+            self.map.insert(accept.clone(), self_addr.clone());
+            // Remove value from pending_connections list
+            self.pending_connections.retain(|x| x != &accept_str);
+        }
+
+        Ok(())
+    }
+
+    async fn handle_route(&mut self, ctx: &Context, mut msg: LocalMessage) -> Result<()> {
+        trace!("Route request: {:?}", msg.transport().onward_route.next());
+
+        // Get the next hop
+        let onward = msg.transport().onward_route.next()?;
+
+        let next;
+        // Look up the connection worker responsible
+        if let Some(n) = self.map.get(&onward) {
+            // Connection already exists
+            next = n;
+        } else {
+            // No existing connection
+            let peer_str;
+            if let Ok(s) = String::from_utf8(onward.deref().clone()) {
+                peer_str = s;
+            } else {
+                return Err(TransportError::UnknownRoute.into());
+            }
+
+            // TODO: Check if this is the hostname and we have existing/pending connection to this IP
+
+            let peer_addr_str = format!("{}#{}", self.addr_id, &peer_str);
+            if self.pending_connections.contains(&peer_addr_str) {
+                // We already trying to connect to this address - Requeue the message
+                ctx.forward(msg).await?;
+            } else if self.allow_auto_connection {
+                // Create connection
+                self.pending_connections.push(peer_addr_str);
+                let handle = self.new_handler(ctx).await?;
+                let _ = handle.connect(peer_str).await?;
+                // Requeue the message
+                ctx.forward(msg).await?;
+            } else {
+                return Err(TransportError::UnknownRoute.into());
+            }
+
+            return Ok(());
+        }
+
+        let _ = msg.transport_mut().onward_route.step()?;
+        // Modify the transport message route
+        msg.transport_mut()
+            .onward_route
+            .modify()
+            .prepend(next.clone());
+
+        // Send the transport message to the connection worker
+        ctx.send(next.clone(), msg).await?;
+
+        Ok(())
+    }
+}
+
+#[worker]
+impl Worker for Router {
+    type Message = RouterMessage;
+    type Context = Context;
+
+    async fn shutdown(&mut self, ctx: &mut Self::Context) -> Result<()> {
+        for (_accepts, self_addr) in self.map.iter() {
+            ctx.stop_worker(self_addr.clone()).await?;
```

Comment:
We probably should ignore errors here. We should keep in mind that any of the workers could be shut down by the node